packages/workflow/src/utils.ts

```diff
@@ -1,37 +1,41 @@
-/* eslint-disable @typescript-eslint/no-explicit-any, @typescript-eslint/no-unsafe-assignment, @typescript-eslint/no-unsafe-member-access, @typescript-eslint/no-unsafe-return */
-
-type Serializable = { toJSON?: () => string };
-
-export const deepCopy = <T>(source: T, hash = new WeakMap(), path = ''): T => {
+/* eslint-disable @typescript-eslint/no-explicit-any, @typescript-eslint/no-unsafe-assignment, @typescript-eslint/no-unsafe-member-access, @typescript-eslint/no-unsafe-return, @typescript-eslint/no-unsafe-argument */
+type Primitives = string | number | boolean | bigint | symbol | null | undefined;
+export const deepCopy = <T extends ((object | Date) & { toJSON?: () => string }) | Primitives>(
+	source: T,
+	hash = new WeakMap(),
+	path = '',
+): T => {
 	let clone: any;
+	let i: any;
 	const hasOwnProp = Object.prototype.hasOwnProperty.bind(source);
 	// Primitives & Null & Function
-	if (typeof source !== 'object' || source === null || source instanceof Function) {
+	if (typeof source !== 'object' || source === null || typeof source === 'function') {
 		return source;
 	}
-	// Date and other Serializable objects
-	const toJSON = (source as Serializable).toJSON;
-	if (typeof toJSON === 'function') {
-		// eslint-disable-next-line @typescript-eslint/no-unsafe-call
-		return toJSON.call(source) as T;
+	// TODO: remove this when other code parts not expecting objects with `.toJSON` method called
+	if (typeof source.toJSON === 'function') {
+		return source.toJSON() as T;
 	}
-	// Break any cyclic dependencies
 	if (hash.has(source)) {
 		return hash.get(source);
 	}
+	// Date
+	if (source instanceof Date) {
```

Comment:
this is already handled by the `toJSON` block because of `Date.prototype.toJSON()`

Reply:
thanks, removed